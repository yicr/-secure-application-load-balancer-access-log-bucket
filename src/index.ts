import { SecureBucket, SecureBucketEncryption } from '@yicr/secure-bucket';
import * as cdk from 'aws-cdk-lib';
import * as iam from 'aws-cdk-lib/aws-iam';
import * as s3 from 'aws-cdk-lib/aws-s3';
import { Construct } from 'constructs';

export interface SecureApplicationLoadBalancerAccessLogBucketProps {
  readonly bucketName?: string;
}

export class SecureApplicationLoadBalancerAccessLogBucket extends SecureBucket {

  constructor(scope: Construct, id: string, props?: SecureApplicationLoadBalancerAccessLogBucketProps) {
    super(scope, id, {
      bucketName: props?.bucketName,
      encryption: SecureBucketEncryption.KMS_MANAGED,
    });

    const account = cdk.Stack.of(this).account;
    const region = cdk.Stack.of(this).region;

    // ðŸ‘‡Add Lifecycle rule.
    this.addLifecycleRule({
      id: 'ArchiveStepLifeCycle',
      enabled: true,
      transitions: [
        {
          storageClass: s3.StorageClass.INFREQUENT_ACCESS,
          transitionAfter: cdk.Duration.days(60),
        },
        {
          storageClass: s3.StorageClass.INTELLIGENT_TIERING,
          transitionAfter: cdk.Duration.days(120),
        },
        {
          storageClass: s3.StorageClass.GLACIER,
          transitionAfter: cdk.Duration.days(180),
        },
        {
          storageClass: s3.StorageClass.DEEP_ARCHIVE,
          transitionAfter: cdk.Duration.days(360),
        },
      ],
    });

    // ðŸ‘‡Allow Region ALB Account
    this.addToResourcePolicy(new iam.PolicyStatement({
      effect: iam.Effect.ALLOW,
      actions: [
        's3:PutObject',
      ],
      principals: [
        new iam.AccountPrincipal((() => {
          // https://docs.aws.amazon.com/elasticloadbalancing/latest/application/enable-access-logging.html
          switch (region) {
            case 'us-east-1': // US East (N. Virginia)
              return '127311923021';
            case 'us-east-2': // US East (Ohio)
              return '033677994240';
            case 'us-west-1': // US West (N. California)
              return '027434742980';
            case 'us-west-2': // US West (Oregon)
              return '797873946194';
            case 'us-gov-west-1': // AWS GovCloud (US-West)
              return '048591011584';
            case 'us-gov-east-1': // WS GovCloud (US-East)
              return '190560391635';
            case 'ca-central-1': // Canada (Central)
              return '985666609251';
            case 'eu-central-1': // Europe (Frankfurt)
              return '054676820928';
            case 'eu-north-1': // Europe (Stockholm)
              return '897822967062';
            case 'eu-west-1': // Europe (Ireland)
              return '156460612806';
            case 'eu-west-2': // Europe (London)
              return '652711504416';
            case 'eu-west-3': // Europe (Paris)
              return '009996457667';
            case 'eu-south-1': // Europe (Milan)
              return '635631232127';
            case 'ap-northeast-1': // Asia Pacific (Tokyo)
              return '582318560864';
            case 'ap-northeast-2': // Asia Pacific (Seoul)
              return '600734575887';
            case 'ap-northeast-3': // Asia Pacific (Osaka)
              return '383597477331';
            case 'ap-southeast-1': // Asia Pacific (Singapore)
              return '114774131450';
            case 'ap-southeast-2': // Asia Pacific (Sydney)
              return '783225319266';
            case 'ap-southeast-3': // Asia Pacific (Jakarta)
              return '589379963580';
            case 'ap-south-1': // Asia Pacific (Mumbai)
              return '718504428378';
            case 'ap-east-1': // Asia Pacific (Hong Kong)
              return '754344448648';
            case 'af-south-1': // Africa (Cape Town)
              return '098369216593';
            // case 'cn-north-1': // China (Beijing)
            //   return '';
            // case 'cn-northwest-1': // China (Ningxia)
            //   return '';
            case 'me-south-1': // Middle East (Bahrain)
              return '076674570225';
            case 'sa-east-1': // South America (SÃ£o Paulo)
              return '507241528517';
            default:
              throw new Error('Unknown region or Undefined region.');
          }
        })()),
      ],
      resources: [`${this.bucketArn}/AWSLogs/${account}/*`],
    }));

  }
}