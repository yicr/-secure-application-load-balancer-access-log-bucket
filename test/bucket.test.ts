import { App, Stack } from 'aws-cdk-lib';
import { Match, Template } from 'aws-cdk-lib/assertions';
import * as s3 from 'aws-cdk-lib/aws-s3';
import { SecureApplicationLoadBalancerAccessLogBucket } from '../src';

describe('SecureApplicationLoadBalancerAccessLogBucket Testing', () => {

  const app = new App();
  const stack = new Stack(app, 'TestingStack', {
    env: {
      account: '123456789012',
      region: 'us-east-1',
    },
  });

  const bucket = new SecureApplicationLoadBalancerAccessLogBucket(stack, 'SecureApplicationLoadBalancerAccessLogBucket', {
    bucketName: 'example-alb-access-log-bucket',
    account: '123456789012',
    region: 'us-east-1',
  });

  it('Is Bucket', () => {
    expect(bucket).toBeInstanceOf(s3.Bucket);
  });

  const template = Template.fromStack(stack);

  it('Should have encryption', () => {
    template.hasResourceProperties('AWS::S3::Bucket', {
      BucketEncryption: Match.objectEquals({
        ServerSideEncryptionConfiguration: [
          {
            ServerSideEncryptionByDefault: {
              SSEAlgorithm: 'aws:kms',
            },
          },
        ],
      }),
    });
  });

  it('Should versioning enabled', () => {
    template.hasResourceProperties('AWS::S3::Bucket', {
      VersioningConfiguration: Match.objectEquals({
        Status: 'Enabled',
      }),
    });
  });

  it('Should exist lifecycle', () => {
    template.hasResourceProperties('AWS::S3::Bucket', {
      LifecycleConfiguration: {
        Rules: Match.arrayEquals([
          Match.objectEquals({
            Id: 'ArchiveStepLifeCycle',
            Status: 'Enabled',
            Transitions: Match.arrayEquals([
              Match.objectEquals({
                StorageClass: 'STANDARD_IA',
                TransitionInDays: 60,
              }),
              Match.objectEquals({
                StorageClass: 'INTELLIGENT_TIERING',
                TransitionInDays: 120,
              }),
              Match.objectEquals({
                StorageClass: 'GLACIER',
                TransitionInDays: 180,
              }),
              Match.objectEquals({
                StorageClass: 'DEEP_ARCHIVE',
                TransitionInDays: 360,
              }),
            ]),
          }),
        ]),
      },
    });
  });

  it('Should enforce SSL', () => {
    template.hasResourceProperties('AWS::S3::BucketPolicy', {
      Bucket: {
        Ref: Match.stringLikeRegexp('SecureApplicationLoadBalancerAccessLogBucket'),
      },
      PolicyDocument: {
        Version: '2012-10-17',
        Statement: Match.arrayWith([
          Match.objectLike({
            Action: 's3:*',
            Condition: {
              Bool: {
                'aws:SecureTransport': 'false',
              },
            },
            Effect: 'Deny',
            Principal: {
              AWS: '*',
            },
          }),
        ]),
      },
    });
  });

  it('Should match snapshot', () => {
    expect(template.toJSON()).toMatchSnapshot();
  });
});

describe('Region SecureApplicationLoadBalancerAccessLogBucket Testing', () => {
  const cases = [
    ['us-east-1', '127311923021'], // US East (N. Virginia)
    ['us-east-2', '033677994240'], // US East (Ohio)
    ['us-west-1', '027434742980'], // US West (N. California)
    ['us-west-2', '797873946194'], // US West (Oregon)
    ['us-gov-west-1', '048591011584'], // AWS GovCloud (US-West)
    ['us-gov-east-1', '190560391635'], // WS GovCloud (US-East)
    ['ca-central-1', '985666609251'], // Canada (Central)
    ['eu-central-1', '054676820928'], // Europe (Frankfurt)
    ['eu-north-1', '897822967062'], // Europe (Stockholm)
    ['eu-west-1', '156460612806'], // Europe (Ireland)
    ['eu-west-2', '652711504416'], // Europe (London)
    ['eu-west-3', '009996457667'], // Europe (Paris)
    ['eu-south-1', '635631232127'], // Europe (Milan)
    ['ap-northeast-1', '582318560864'], // Asia Pacific (Tokyo)
    ['ap-northeast-2', '600734575887'], // Asia Pacific (Seoul)
    ['ap-northeast-3', '383597477331'], // Asia Pacific (Osaka)
    ['ap-southeast-1', '114774131450'], // Asia Pacific (Singapore)
    ['ap-southeast-2', '783225319266'], // Asia Pacific (Sydney)
    ['ap-southeast-3', '589379963580'], // Asia Pacific (Jakarta)
    ['ap-south-1', '718504428378'], // Asia Pacific (Mumbai)
    ['ap-east-1', '754344448648'], // Asia Pacific (Hong Kong)
    ['af-south-1', '098369216593'], // Africa (Cape Town)
    ['me-south-1', '076674570225'], // Middle East (Bahrain)
    ['sa-east-1', '507241528517'], // South America (SÃ£o Paulo)
  ];
  it.each(cases)('Should match region %p = account(%p)', (regionCode, albAwsAccount) => {
    const app = new App();
    const stack = new Stack(app, `TestingStack${regionCode}`, {
      env: {
        account: '123456789012',
        region: 'us-east-1',
      },
    });
    new SecureApplicationLoadBalancerAccessLogBucket(stack, 'SecureApplicationLoadBalancerAccessLogBucket', {
      bucketName: 'example-alb-access-log-bucket',
      account: '123456789012',
      region: regionCode,
    });
    const template = Template.fromStack(stack);
    template.hasResourceProperties('AWS::S3::BucketPolicy', {
      Bucket: {
        Ref: Match.stringLikeRegexp('SecureApplicationLoadBalancerAccessLogBucket'),
      },
      PolicyDocument: {
        Version: '2012-10-17',
        Statement: Match.arrayWith([
          Match.objectEquals({
            Effect: 'Allow',
            Action: 's3:PutObject',
            Principal: {
              AWS: {
                'Fn::Join': Match.arrayWith([
                  '',
                  [
                    'arn:',
                    {
                      Ref: 'AWS::Partition',
                    },
                    `:iam::${albAwsAccount}:root`,
                  ],
                ]),
              },
            },
            Resource: {
              'Fn::Join': Match.arrayWith([
                '',
                [
                  {
                    'Fn::GetAtt': Match.arrayWith([
                      Match.stringLikeRegexp('SecureApplicationLoadBalancerAccessLogBucket'),
                      'Arn',
                    ]),
                  },
                  '/AWSLogs/123456789012/*',
                ],
              ]),
            },
          }),
        ]),
      },
    });
  });

  it('Should unknown region error', () => {
    const app = new App();
    const stack = new Stack(app, 'ErrorTestingStack', {
      env: {
        account: '123456789012',
        region: 'us-east-1',
      },
    });
    const bucket = () => {
      new SecureApplicationLoadBalancerAccessLogBucket(stack, 'SecureApplicationLoadBalancerAccessLogBucket', {
        bucketName: 'example-alb-access-log-bucket',
        account: '123456789012',
        region: 'xx-xxx-1',
      });
    };
    expect(bucket).toThrowError(Error);
  });
});