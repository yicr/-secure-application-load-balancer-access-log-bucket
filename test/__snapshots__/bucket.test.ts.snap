// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SecureApplicationLoadBalancerAccessLogBucket Testing Should match snapshot 1`] = `
Object {
  "Parameters": Object {
    "BootstrapVersion": Object {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": Object {
    "SecureApplicationLoadBalancerAccessLogBucket6DB3A9E7": Object {
      "DeletionPolicy": "Retain",
      "Properties": Object {
        "AccessControl": "Private",
        "BucketEncryption": Object {
          "ServerSideEncryptionConfiguration": Array [
            Object {
              "ServerSideEncryptionByDefault": Object {
                "SSEAlgorithm": "aws:kms",
              },
            },
          ],
        },
        "LifecycleConfiguration": Object {
          "Rules": Array [
            Object {
              "Id": "ArchiveStepLifeCycle",
              "Status": "Enabled",
              "Transitions": Array [
                Object {
                  "StorageClass": "STANDARD_IA",
                  "TransitionInDays": 60,
                },
                Object {
                  "StorageClass": "INTELLIGENT_TIERING",
                  "TransitionInDays": 120,
                },
                Object {
                  "StorageClass": "GLACIER",
                  "TransitionInDays": 180,
                },
                Object {
                  "StorageClass": "DEEP_ARCHIVE",
                  "TransitionInDays": 360,
                },
              ],
            },
          ],
        },
        "PublicAccessBlockConfiguration": Object {
          "BlockPublicAcls": true,
          "BlockPublicPolicy": true,
          "IgnorePublicAcls": true,
          "RestrictPublicBuckets": true,
        },
        "VersioningConfiguration": Object {
          "Status": "Enabled",
        },
      },
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Retain",
    },
    "SecureApplicationLoadBalancerAccessLogBucketPolicyDCDADB99": Object {
      "Properties": Object {
        "Bucket": Object {
          "Ref": "SecureApplicationLoadBalancerAccessLogBucket6DB3A9E7",
        },
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "s3:*",
              "Condition": Object {
                "Bool": Object {
                  "aws:SecureTransport": "false",
                },
              },
              "Effect": "Deny",
              "Principal": Object {
                "AWS": "*",
              },
              "Resource": Array [
                Object {
                  "Fn::GetAtt": Array [
                    "SecureApplicationLoadBalancerAccessLogBucket6DB3A9E7",
                    "Arn",
                  ],
                },
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      Object {
                        "Fn::GetAtt": Array [
                          "SecureApplicationLoadBalancerAccessLogBucket6DB3A9E7",
                          "Arn",
                        ],
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
            Object {
              "Action": "s3:PutObject",
              "Effect": "Allow",
              "Principal": Object {
                "AWS": Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "arn:",
                      Object {
                        "Ref": "AWS::Partition",
                      },
                      ":iam::127311923021:root",
                    ],
                  ],
                },
              },
              "Resource": Object {
                "Fn::Join": Array [
                  "",
                  Array [
                    Object {
                      "Fn::GetAtt": Array [
                        "SecureApplicationLoadBalancerAccessLogBucket6DB3A9E7",
                        "Arn",
                      ],
                    },
                    "/AWSLogs/123456789012/*",
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::S3::BucketPolicy",
    },
  },
  "Rules": Object {
    "CheckBootstrapVersion": Object {
      "Assertions": Array [
        Object {
          "Assert": Object {
            "Fn::Not": Array [
              Object {
                "Fn::Contains": Array [
                  Array [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  Object {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
